pkg/translator/translator.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright © 2022 Kubernetes Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package translator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	networkingv1 "k8s.io/api/networking/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/yaml"
+)
+
+type Translator struct {
+	Options
+}
+
+func New(options Options) *Translator {
+	return &Translator{
+		options,
+	}
+}
+
+type Options struct {
+	Writer       io.Writer
+	Mode         string
+	FilePath     string
+	OutputType   string
+	ResourceType string
+	Provider     IngressProvider
+}
+
+func WithOptions(w io.Writer, mode, filePath, outputType, resourceType, provider string) Options {
+	return Options{
+		Writer:       w,
+		FilePath:     filePath,
+		OutputType:   outputType,
+		ResourceType: resourceType,
+		Mode:         mode,
+		Provider:     IngressProvider(provider),
+	}
+}
+
+func (t Translator) Run() error {
+	if err := t.validate(); err != nil {
+		return err
+	}
+
+	var (
+		result *Result
+		errs   []error
+	)
+
+	ingresses, err := t.ConstructIngresses(t.Mode, t.FilePath)
+	if err != nil {
+		return err
+	}
+
+	result, errs = t.Convert(t.Provider, ingresses)
+	if len(errs) != 0 {
+		fmt.Fprintln(t.Writer, errs)
+	}
+
+	data, err := t.Marshal(*result, t.OutputType, t.ResourceType)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintln(t.Writer, string(data))
+
+	return nil
+}
+
+func (t Translator) validate() error {
+	if !isValidResourceType(GWAPIResourceType(t.ResourceType)) {
+		return fmt.Errorf("%s is not a valid output type. %s", t.ResourceType, GetValidResourceTypesStr())
+	}
+
+	return nil
+}
+
+func (t Translator) ConstructIngresses(mode, file string) ([]networkingv1.Ingress, error) {
```

Comment:
It would be helpful to have comments describing what this and other exported functions are doing. In this case I'd be tempted to call this `ReadIngresses` because `Construct` makes me think we're building new things, not just reading existing things.